src/connection.cpp

```diff
@@ -168,7 +168,12 @@ static int connSocketWrite(connection *conn, const void *data, size_t data_len)
     int ret = write(conn->fd, data, data_len);
     if (ret < 0 && errno != EAGAIN) {
         conn->last_errno = errno;
-        conn->state.store(CONN_STATE_ERROR, std::memory_order_relaxed);
+
+        /* Don't overwrite the state of a connection that is not already
+         * connected, not to mess with handler callbacks.
+         */
+        ConnectionState expected = CONN_STATE_CONNECTED;
+        conn->state.compare_exchange_strong(expected, CONN_STATE_ERROR);
```

Comment:
Why are we modifying this to only sometimes change the value (and not listen to the result of whether it changed or not)?

Reply:
Maybe I'm misunderstanding how compare_exchange works - I hadn't come across it before - but this does the same as the code I replaced. It compares conn->state with expected, and if they match, it replaces conn->state with CONN_STATE_ERROR.